### ISSUE TITLE
Ratio between 'Actual rows' and 'Rows Removed by Filter'

### ISSUE THREAD

**Issue by Krysztophe** [dalibo/pev2#98]
A very small ratio between returned rows and the total number of scanned rows (not shown in the plan but available as Actual rows + Rows Removed by Filter) is an indication that an index might be a good idea, so an hint should  appear somewhere.

In a 1st time, add a percentage of filtered on the same line as "Rows removed by Filter" (something like ">99.99%" in the following example), and "Actual rows" ("<0.01%" here).
Rounding: 2 digits ("25%") is fine, but when you get very close to 0% or 99% you should not round to 0% or 100% (except if 0 or all lines are filtered of course).
I'm wondering if it should be "99.99.% of the scanned rows" to avoid confusion with "99.99% of the table" (which would be another )

(BTW : add a  "Total scanned rows" line ?)

In a 2nd time, add something in the same vein as "overestimated rows by ..." in the default view of the node. I don't know which color to use and suggest that an index would be useful, there is no real rule there (orange for 80% of removed lines ? 90% ?)

1st example (without parallelism) : 1 line is returned among 1 million.
```
 select * from t_un_million_int where i+0 = 500000 ;
```
```
 Seq Scan on t_un_million_int  (cost=0.00..19425.00 rows=5000 width=4) (actual time=66.075..101.704 rows=1 loops=1)
   Filter: ((i + 0) = 500000)
   Rows Removed by Filter: 999999
 Planning Time: 0.153 ms
 Execution Time: 101.755 ms
(5 lignes)
```
Same query, with parallelism: Removed lines are to be multiplied by 3 total workers, and Actual rows should come from Gather node, not from the parallel scans (average is zero):
```
 Gather  (cost=1000.00..12175.00 rows=5000 width=4) (actual time=124.521..127.080 rows=1 loops=1)
   Workers Planned: 2
   Workers Launched: 2
   ->  Parallel Seq Scan on t_un_million_int  (cost=0.00..10675.00 rows=2083 width=4) (actual time=97.027..111.253 rows=0 loops=3)
         Filter: ((i + 0) = 500000)
         Rows Removed by Filter: 333333
 Planning Time: 0.179 ms
 Execution Time: 127.128 ms
(8 lignes)
```

2nd query, with intermediate index: it should show "Rows Removed by Filter: 95% of scanned rows"  (20/21)
```
explain (analyze) select * from t_un_million_int where i between 499990 and 500010 and i+0 =500000  ;
                                                                 QUERY PLAN                                                                  
---------------------------------------------------------------------------------------------------------------------------------------------
 Index Only Scan using t_un_million_int_pkey on t_un_million_int  (cost=0.42..1.93 rows=1 width=4) (actual time=0.058..0.065 rows=1 loops=1)
   Index Cond: ((i >= 499990) AND (i <= 500010))
   Filter: ((i + 0) = 500000)
   Rows Removed by Filter: 20
   Heap Fetches: 0
 Planning Time: 0.228 ms
 Execution Time: 0.126 ms
(7 lignes)
```

**Comment by pgiraud** [comment]
It looks like we can have "Rows Removed by XXX" where XXX can be either:
 - "Filter",
 - "Join Filter",
 - "Index Recheck".

There are cases where we have 2 of those counts (for example `Filter` + `Index Recheck`).
Should I sum them up to get the total of discarded rows?

**Comment by pgiraud** [comment]
It looks like we can have "Rows Removed by XXX" where XXX can be either:
 - "Filter",
 - "Join Filter",
 - "Index Recheck".

There are cases where we have 2 of those counts (for example `Filter` + `Index Recheck`).
Should I sum them up to get the total of discarded rows?

Or should I treat the different cases differently and only concentrate on "by Filter" for now?

**Comment by pgiraud** [comment]
`by Join Filter` and `by Index Recheck` have different meaning. I will not take them into account for now.

**Comment by pgiraud** [comment]
Here are several alternatives to show the information. I personnaly prefer the last one.

![Capture du 2020-02-28 09-08-56](https://user-images.githubusercontent.com/319774/75526209-9fca4a80-5a11-11ea-8c80-1829a16edbfe.png)
![Capture du 2020-02-28 09-36-29](https://user-images.githubusercontent.com/319774/75526212-a062e100-5a11-11ea-89dc-0d48e76edce4.png)
![Capture du 2020-02-28 09-40-15](https://user-images.githubusercontent.com/319774/75526213-a0fb7780-5a11-11ea-9cb3-e832d302329f.png)

**Comment by Krysztophe** [comment]
I think that the ratio is the most important, so it should appear in any case. In this example it is obvious that 190165 >> 522 , but is not obvious at first sight that 91234567899 >> 9123456799 (or you must round to have 91 billions vs 9 billions).

Keep the color on the 190165  if you prefer (it's only there to catch the eye after all). 

So I'd vote for the 1st or a mix with the others:
```
Rows removed by FIlter : **190165** out of 190687 scanned rows (99%), 522 returned
```
But this is maybe too much on one line.

**Comment by Krysztophe** [comment]
About "Rows Removed by Join Filter'': I think this is a similar problem, an indication that some filtering occur that an index may perhaps solve.

''Index Recheck'' is another beast, this is more a problem of index efficiency (a bit like the _Heap fetches_ in #99), that may be interesting too.

**Comment by pgiraud** [comment]
Here's an additional example showing that the number of rows actually returned should be taken in the direct parent not the gather node.

```
explain (analyze) select sum(i) from t_un_million_int where i+0 = 500000 ;
                                                                  QUERY PLAN
----------------------------------------------------------------------------------------------------------------------------------------------
 Finalize Aggregate  (cost=11680.42..11680.43 rows=1 width=8) (actual time=56.924..56.925 rows=1 loops=1)
   ->  Gather  (cost=11680.21..11680.42 rows=2 width=8) (actual time=56.657..63.498 rows=3 loops=1)
         Workers Planned: 2
         Workers Launched: 2
         ->  Partial Aggregate  (cost=10680.21..10680.22 rows=1 width=8) (actual time=49.794..49.795 rows=1 loops=3)
               ->  Parallel Seq Scan on t_un_million_int  (cost=0.00..10675.00 rows=2083 width=4) (actual time=42.558..49.772 rows=0 loops=3)
                     Filter: ((i + 0) = 500000)
                     Rows Removed by Filter: 333333
 Planning Time: 0.129 ms
 Execution Time: 63.581 ms
(10 lignes)
```

**Comment by pgiraud** [comment]
After a deeper investigation I think it's a bad idea to take parent nodes into account to count the number of actual rows. First of all, the number of actual rows is indeed the average number of rows for each loop. And not only in the case of parallel nodes. It has to be multiplied by the number of loops in order to get the number of rows returned by the node. It is imprecise though since the average value is rounded. In some cases the average value is 0. But it's impossible to know what's the actual value for the node itself.
Nevertheless, it is in my opinion not very safe to guess the value from the parent node. Because it may be very different from one node type to another.

So I decided to use the values that are given in the plan without trying to guess more. In some cases, the UI will show "100% of rows removed" which may not be very accurate. When it's not accurate (because of several loops), the UI will show that the values are approximated.

![Capture du 2020-04-17 17-01-18](https://user-images.githubusercontent.com/319774/79583269-2d9aeb80-80cd-11ea-903e-4f2cb14e9da3.png)


As suggested by @ioguix, I also took the duration into account to show and highlight the "rows removed ratio". The ratio itself is not very relevant. It can be very high but with almost no impact on the query duration. It is relevant if the relative duration is high as well.
For example, if a node takes 1% of the whole execution time, we don't really care if 90% of the rows were removed for this node. However, 50% of rows removed in a node taking 95% of the execution time must be taken into account.
